flashlight/fl/nn/modules/Module.h

```diff
@@ -122,6 +129,13 @@ class Module {
    */
   std::vector<Variable> operator()(const std::vector<Variable>& inputs);
 
+  /**
+   * Clone the module via deep copy of its parameters and members.
+   *
+   * @return A shared pointer of the cloned module.
+   */
+  virtual std::shared_ptr<Module> clone() const = 0;
```

Comment:
Why have `clone()` return a shared pointer instead of an instance? Returning an instance makes things simpler:

```cpp
Module m;
auto p = std::make_shared<T>(std::move(m.clone()));
```
gives a shared pointer, but it also maintains optionality for the user to do whatever they want with the returned instance, i.e. keep it stack allocated, dump it in a `unique_ptr`, etc.

concretely, that would be:
```cpp
virtual Module clone() const = 0
```

Reply:
The standard does not allow an [abstract class](https://en.cppreference.com/w/cpp/language/abstract_class) to be used as a return type unless it is a pointer or reference. Given `shared_ptr` is already used for storing `Modules` in `Containers`, I opted to use `shared_ptr`.

I think there is a strong argument for returning a `unique_ptr` instead, but it would mean changing the implementation of `Container` and any classes which inherit it. Additionally there may be a bunch of flow on effects which need to be considered. However that's beyond the scope of this PR, so it's probably best to stick with `shared_ptr` for now.